File: LeetCode/Study Plan/Practice Problems/Easy/Palindrome Linked List/Go/palindromeLinkedList.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func New(val int) *ListNode { return &ListNode{Val: val, Next: nil} }

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	return checkPalindrome(head, &head)
}

func checkPalindrome(temp *ListNode, head **ListNode) bool {
	if temp != nil {
		if !checkPalindrome(temp.Next, head) {
			return false
		}
		if (*head).Val != temp.Val {
			return false
		}
		*head = (*head).Next
	}
	return true
}

func main() {
	head := New(1)
	head.Next = New(2)
	head.Next.Next = New(2)
	head.Next.Next.Next = New(1)
	// fmt.Println(is_palindrome(head))
	temp := head
	fmt.Println(checkPalindrome(head, &temp))
}

File: LeetCode/Easy/Palindrome Linked List/Java/palindromeLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode headGlobal;
    public boolean isPalindrome(ListNode head) {
        headGlobal = head;
        return checkPalindrome(head);
    }

    public boolean checkPalindrome(ListNode head) {
        if (head != null) {
            if (!checkPalindrome(head.next)) {
                return false;
            }
            if (head.val != headGlobal.val) {
                return false;
            }
            headGlobal = headGlobal.next;
        }
        return true;
    }
}
